File: 36-valid-sudoku/valid-sudoku.cpp
class Solution {
public:
    bool is_valid(int s, int e, vector<vector<char>>& board) {
        map<char, int> mp;
        for (int i = s; i < s + 3; i++) {
            for (int j = e; j < e + 3; j++) {
                if (board[i][j] == '.') continue;
                if (mp.find(board[i][j]) != mp.end()) return false;
                else mp[board[i][j]]++;
            }
        }
        return true;
    }
    bool isValidSudoku(vector<vector<char>>& board) {
        int n = board.size(); 
        int m = board[0].size(); 
        map<char, int> mp;
        for (int i = 0; i < n; i++) {
            mp.clear();
            for (int j = 0; j < m; j++) {
                if (board[i][j] == '.') continue;
                if (mp.find(board[i][j]) != mp.end()) return false;
                else mp[board[i][j]]++;
            }
        }
        for (int j = 0; j < m; j++) {
            mp.clear();
            for (int i = 0; i < n; i++) {
                if (board[i][j] == '.') continue;
                if (mp.find(board[i][j]) != mp.end()) return false;
                else mp[board[i][j]]++;
            }
        }
        for (int i = 0; i < n; i += 3) {
            for (int j = 0; j < m; j += 3) {
                if (!is_valid(i, j, board)) return false;
            }
        }

        return true;
    }
};

File: 36-valid-sudoku/valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        for(int i=0; i<9; i++) {
            HashSet<Character> hs = new HashSet<>();
            for(int j=0; j<9; j++) {
                char ch = board[i][j];
                if(ch == '.') continue;
                if(hs.contains(ch)) return false;
                hs.add(ch);
            }
        }
        for(int i=0; i<9; i++) {
            HashSet<Character> hs = new HashSet<>();
            for(int j=0; j<9; j++) {
                char ch = board[j][i];
                if(ch == '.') continue;
                if(hs.contains(ch)) return false;
                hs.add(ch);
            }
        }
        for(int i=0; i<9; i+=3) {
            for(int j=0; j<9; j+=3) {
                if(!check(board,i,j)) return false;
            }
        }
        return true;
    }
    public static boolean check(char[][] board, int r ,int c) {
        HashSet<Character> hs = new HashSet<>();
        for(int i=r; i<r+3; i++) {
            for(int j=c; j<c+3;j++) {
                char ch = board[i][j];
                if(ch == '.') continue;
                if(hs.contains(ch)) return false;
                hs.add(ch);
            }
        }
        return true;
    }
}